### PR TITLE
es_objects plugin

### PR DESCRIPTION
**Pull request by oxarbitrage**
Plugin will store certain objects from the blockchain to ES database. Can be easy extended to support more objects. Initial idea was to store proposal objects for https://github.com/bitshares/bitshares-core/issues/115

Instead of loading the database with more data i decided to store the proposal_objects to ES where no object is deleted while we leave the core to keep with the reduced active proposal list.

I noticed then that we could store account data so we can go over https://github.com/bitshares/bitshares-core/issues/452 without loading more the blockchain database with a working referral index. By saving the account data we can index by referrer with ease(ill post some samples). Another advantage is in the case we want to get the proxies, currently we have to loop throw all the accounts and checking the `options.vote_as` field. Very intensive.

Asset and balance objects were also added to the plugin to solve similar issues and make the searching inside them easier for the client apps. 

Runs good among the original `elasticsearch` plugin for operations.A starting command with the 2 plugins can be something as:

`programs/witness_node/witness_node --data-dir data/my-blockprod --rpc-endpoint "127.0.0.1:8090" --plugins "elasticsearch es_objects" --es-objects-proposals true --es-objects-accounts true --es-objects-assets true --es-objects-balances true`

Please let me know what do you guys think.

### CODE REVIEW

File: libraries/plugins/es_objects/es_objects.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+      es_objects_plugin& _self;
+      std::string _es_objects_elasticsearch_url = "http://localhost:9200/";
+      uint32_t _es_objects_bulk_replay = 1000;
+      uint32_t _es_objects_bulk_sync = 2;
+      bool _es_objects_proposals = true;
+      bool _es_objects_accounts = true;
+      bool _es_objects_assets = true;
+      bool _es_objects_balances = true;
+      bool _es_objects_logs = true;
+      CURL *curl; // curl handler
```
Comment (abitmore) [inline libraries/plugins/es_objects/es_objects.cpp]: Do we want to use a new curl handler for each plugin?

File: libraries/plugins/es_objects/es_objects.cpp [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+      else if(value.space() == 1 && value.type() == 3 && _es_objects_assets) {
+         auto obj = db.find_object(value);
+         auto a = static_cast<const asset_object*>(obj);
+         PrepareAsset(a);
+      }
+      else if(value.space() == 1 && value.type() == 15 && _es_objects_balances) {
+         auto obj = db.find_object(value);
+         auto b = static_cast<const balance_object*>(obj);
+         PrepareBalance(b);
+      }
```
Comment (abitmore) [inline libraries/plugins/es_objects/es_objects.cpp]: It's best if we can also store account statistics, dynamic asset data, bitasset data and etc along with the main objects.

File: libraries/plugins/es_objects/include/graphene/es_objects/es_objects.hpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+// conflicting SPACE_ID assignments can be compiled into the
+// same binary (by simply re-assigning some of the conflicting #defined
+// SPACE_ID's in a build script).
+//
+// Assignment of SPACE_ID's cannot be done at run-time because
+// various template automagic depends on them being known at compile
+// time.
+//
+#ifndef PROPOSAL_SPACE_ID
+#define PROPOSAL_SPACE_ID 6
```
Comment (abitmore) [inline libraries/plugins/es_objects/include/graphene/es_objects/es_objects.hpp]: Better use a better name here.
Also I'm not sure if the number `6` is appropriate.

File: libraries/plugins/es_objects/es_objects.cpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+      limit_documents = _es_objects_bulk_sync;
+   else
+      limit_documents = _es_objects_bulk_replay;
+
+   if (curl && bulk.size() >= limit_documents) { // we are in bulk time, ready to add data to elasticsearech
+      SendBulk();
+   }
+
+   for(auto const& value: ids) {
+      if(value.space() == 1 && value.type() == 10 && _es_objects_proposals) {
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: yes, sorry, i was supposed to change that before sending the pull. changed now here https://github.com/bitshares/bitshares-core/pull/500/commits/26c57fda3fb0e317e7e1f49f74613c5adc836d97

File: libraries/plugins/es_objects/es_objects.cpp [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+         auto obj = db.find_object(value);
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: added check for nullptr at https://github.com/bitshares/bitshares-core/pull/500/commits/7b18456c8e924e5180397717b8feccca4d85090f

File: libraries/plugins/es_objects/es_objects.cpp [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+   }
+}
+
+void es_objects_plugin_impl::PrepareProposal(const proposal_object* proposal_object)
+{
+   proposal_struct prop;
+   prop.id = proposal_object->id;
+   prop.expiration_time = proposal_object->expiration_time;
+   prop.review_period_time = proposal_object->review_period_time;
+   prop.proposed_transaction = fc::json::to_string(proposal_object->proposed_transaction);
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: can it be done by adding the proposal objects fields `required_active_approvals`, `available_active_approvals`, `required_owner_approvals`, `available_owner_approvals` and `available_key_approvals` ?
let me know if this will do it so i can add these.

File: libraries/plugins/es_objects/include/graphene/es_objects/es_objects.hpp (hunk 3 above)
Comment (oxarbitrage) [reply libraries/plugins/es_objects/include/graphene/es_objects/es_objects.hpp]: this is something i dont understand fully so maybe you can explain me a bit. the comment states that plugins with the same ID will be added to the same binary so i think maybe is appropriate to have the same ID in the 2 ES plugins as: https://github.com/bitshares/bitshares-core/blob/develop/libraries/plugins/elasticsearch/include/graphene/elasticsearch/elasticsearch_plugin.hpp#L47

but maybe i am wrong, the snapshot plugin from @pmconrad  actually dont define a number(https://github.com/bitshares/bitshares-core/blob/develop/libraries/plugins/snapshot/include/graphene/snapshot/snapshot.hpp) please let me know if we should change this pulgin to number 7. 

in regards to the better name i am not sure what do you mean, there is no name defined here but only the ID. let me know. thanks @abitmore !

Comment (abitmore) [reply libraries/plugins/es_objects/include/graphene/es_objects/es_objects.hpp]: I was referring that you use `es_object` for the plugin name but setting `PROPOSAL` here, it's not good practice. Not about the number.
According to the comment above, there should be a script to change the ID's at building time, but I'm not sure if there is one.
If I understood correctly, the space ID's are used to define a "space" for data storing in the object database. So, if no new object type is defined and need to be stored into object database, no need to set a `XX_SPACE_ID`, as have done with the snapshot plugin.

Comment (oxarbitrage) [reply libraries/plugins/es_objects/include/graphene/es_objects/es_objects.hpp]: oh, sorry about that name, didn't saw it.  it must be there because in a first version i was only going to store proposals. anyways, with your clarification i now removed all that section from the hpp file in commit https://github.com/bitshares/bitshares-core/pull/500/commits/7b453c6a05e4b052b6b04a304e094086b6b2a498 and code stills compiles and works fine.
thanks.

File: libraries/plugins/es_objects/es_objects.cpp [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+   acct.registrar = account_object->registrar;
+   acct.referrer = account_object->referrer;
+   acct.lifetime_referrer = account_object->lifetime_referrer;
+   acct.network_fee_percentage = account_object->network_fee_percentage;
+   acct.lifetime_referrer_fee_percentage = account_object->lifetime_referrer_fee_percentage;
+   acct.referrer_rewards_percentage = account_object->referrer_rewards_percentage;
+   acct.name = account_object->name;
+   acct.owner = fc::json::to_string(account_object->owner.get_addresses());
+   acct.active = fc::json::to_string(account_object->active.get_addresses());
+   acct.voting_account = account_object->options.voting_account;
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: pls explain a bit more this because i don't get it. constructor will be made in the hpp while i don't have this values available there to initialize this way, i am probably missing something, an example may be good. thanks.

File: libraries/plugins/es_objects/es_objects.cpp (hunk 1 above)
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: interesting question, this is debatable at least to me. using the same handler for all requests that goes to the same url is recommended for performance and doing otherwise will be impossible for our purposes in the original plugin(connect and disconnect with each request will be just too slow in a replay).

i made a new handler here because i am not sure how much is too much when they run at the same time. the performance of sending bulks in 5k operation batches were proven to be good enough in a replay scenario while if we add more queries to the same handler we could run into performance issues with more data coming in by new plugins trying to store more data in the database.  for example if first plugin is sending a bulk of 5k operations(10000 lines of text ) and this new plugin is trying to send 1k lines more it will need to wait until the handler is available to do it ending in some delay that i am not sure if it will be a problem in practice. 

basically for that reason a new handler was added and answering the question; yes. the intention was to use a new handler with each plugin so they can be sending data in "parallel". 

i was able to get in sync a node with the 2 plugins running at the same time and it was ok this way, i can make some tests by using the same handler and check that if you think it worth it.

**Comment by oxarbitrage** [comment]
> I slightly tend to merge this plugin into the earlier ES plugin.

After thinking about this i tend to agree and think it could be a good idea to have a single plugin with all the elastic stuff on it. I can close this pull and try to add this to the original if you think it will be better, i don't have any valid argument to don't do it that way neither against multiple plugins so i am not sure.

File: libraries/plugins/es_objects/es_objects.cpp [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+   acct.id = account_object->id;
+   acct.membership_expiration_date = account_object->membership_expiration_date;
+   acct.registrar = account_object->registrar;
+   acct.referrer = account_object->referrer;
+   acct.lifetime_referrer = account_object->lifetime_referrer;
+   acct.network_fee_percentage = account_object->network_fee_percentage;
+   acct.lifetime_referrer_fee_percentage = account_object->lifetime_referrer_fee_percentage;
+   acct.referrer_rewards_percentage = account_object->referrer_rewards_percentage;
+   acct.name = account_object->name;
+   acct.owner = fc::json::to_string(account_object->owner.get_addresses());
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: added in this commit https://github.com/bitshares/bitshares-core/pull/500/commits/3545c22ab806bcb7b2d060da6b6260f4a9eda7c3
i had to add the whole `flat_map` in text form to make it work. this will make the indexing by this fields hard but i think its ok, the real intention is to have something available to relate the `accounts` index with the `balances` that uses an `owner` key.

File: libraries/plugins/es_objects/es_objects.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (abitmore) [reply libraries/plugins/es_objects/es_objects.cpp]: Yes. It worth a try, although I'm not sure whether the final state can be correctly stored with current implementation, because if the proposal has no review period, the object will be removed in the same block when the proposal is approved.

File: libraries/plugins/es_objects/es_objects.cpp (hunk 1 above)
Comment (abitmore) [reply libraries/plugins/es_objects/es_objects.cpp]: Actually I'm not sure, so was asking.
If the 2 plugins are merged into one, there will be only one handler by nature.

File: libraries/plugins/es_objects/es_objects.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (abitmore) [reply libraries/plugins/es_objects/es_objects.cpp]: For example, in hpp file, have
```
struct account_struct {
   string name;
   ...
   account_struct(const account_object* obj) {
      name = obj->name;
      ...
   }
}
```
In cpp file (here), only one line:
```
   account_struct a(pointer_to_the_account_obj);
```

File: libraries/plugins/es_objects/es_objects.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: added on https://github.com/bitshares/bitshares-core/pull/500/commits/fe4e87b8a24800b3b94e589dd61cdc10600b7464

File: libraries/plugins/es_objects/es_objects.cpp [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+void es_objects_plugin_impl::createBulk(std::string type, std::string data, std::string id)
+{
+   bulk.push_back("{ \"index\" : { \"_index\" : \"bitshares-"+type+"\", \"_type\" : \"data\", \"_id\" : "+id+" } }");
+   bulk.push_back(data);
+}
+
+void es_objects_plugin_impl::SendBulk()
+{
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: agree, working on it.

**Comment by pnomolos** [comment]
Hi @oxarbitrage What's the time frame on this do you think?  Thanks!

**Comment by oxarbitrage** [comment]
last commit (https://github.com/bitshares/bitshares-core/pull/500/commits/2bf302d6a855a9f745ae9d9b759e3a53ff2b3517) moves `SendBulk` and `createBulk` common elasticsearch calls to the utilities part of the project. `es_objects` plugin now make uses of them and also original `elasticsearch_plugin` can use them but i will make that in  another pull.

it also add some console output for errors(https://github.com/bitshares/bitshares-core/issues/681), add elasticsearch 6 support and remove unnecessary includes.

**Comment by oxarbitrage** [comment]
> Hi @oxarbitrage What's the time frame on this do you think? Thanks!

I have resumed to work on it to have it included ASAP.

File: libraries/plugins/es_objects/es_objects.cpp [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+      es_objects_plugin_impl(es_objects_plugin& _plugin)
+         : _self( _plugin )
+      {  curl = curl_easy_init(); }
+      virtual ~es_objects_plugin_impl();
+
+      void updateDatabase( const vector<object_id_type>& ids , bool isNew);
+
+      graphene::chain::database& database()
+      {
+         return _self.database();
```
Comment (xeroc) [inline libraries/plugins/es_objects/es_objects.cpp]: What's this for?

File: libraries/plugins/es_objects/es_objects.cpp [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+void es_objects_plugin_impl::updateDatabase( const vector<object_id_type>& ids , bool isNew)
+{
+
+   graphene::chain::database &db = database();
+
+   auto block_time = db.head_block_time();
+
+   // check if we are in replay or in sync and change number of bulk documents accordingly
+   uint32_t limit_documents = 0;
+   if((fc::time_point::now() - block_time) < fc::seconds(30))
```
Comment (xeroc) [inline libraries/plugins/es_objects/es_objects.cpp]: Is this the only way to figure out if we are in replay or normal mode?

File: libraries/plugins/es_objects/es_objects.cpp [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+   if (options.count("es-objects-balances")) {
+      my->_es_objects_balances = options["es-objects-balances"].as<bool>();
+   }
+   if (options.count("es-objects-limit-orders")) {
+      my->_es_objects_limit_orders = options["es-objects-limit-orders"].as<bool>();
+   }
+}
+
+void es_objects_plugin::plugin_startup()
+{
```
Comment (xeroc) [inline libraries/plugins/es_objects/es_objects.cpp]: Can you add a log entry here so people have feedback about es plugin being enabled when starting the backend, please?

**Comment by xeroc** [comment]
Looks great!

File: libraries/plugins/es_objects/es_objects.cpp (hunk 10 above)
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: no need, simplified database call here https://github.com/bitshares/bitshares-core/pull/500/commits/207eb075562f06948e9d62ce693cf800b82b4056 thanks!

File: libraries/plugins/es_objects/es_objects.cpp (hunk 12 above)
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: added at https://github.com/bitshares/bitshares-core/pull/500/commits/109b19f79101a36b7033eee357d3da50751f4b35 thx.

File: libraries/plugins/es_objects/es_objects.cpp (hunk 11 above)
Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: not that i know, this one seems to be effective as it is used in already implemented elasticsearch for account history plugin: https://github.com/bitshares/bitshares-core/blob/master/libraries/plugins/elasticsearch/elasticsearch_plugin.cpp#L198

of course open to ideas to change both.

**Comment by oxarbitrage** [comment]
when saving object bitasset i realized it hits the database hard, even if `feed` and `current_feed` don't change the object is updated with a new date all the time, plus updated when any of the feeds inside `feed` change and finally also updates when `current_feed` change. this last and only this is what we want to track.

so i created a map of bitassets to keep track of the last current_feed string in order to compare when object change and insert records in the database only in that situation.

i didn't synchronized the chain full but it seems to be reasonable. maintaining the map of bitassets will increase RAM but it is better than massively querying the ES database to check if a change in current feed was done.

File: libraries/plugins/es_objects/es_objects.cpp [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+
+   std::string data = fc::json::to_string(acct);
+   prepare = graphene::utilities::createBulk("bitshares-account", data, fc::json::to_string(account_object->id), 0);
+   bulk.insert(bulk.end(), prepare.begin(), prepare.end());
+   prepare.clear();
+}
+
+void es_objects_plugin_impl::PrepareAsset(const asset_object* asset_object)
+{
+   asset_struct asset;
```
Comment (abitmore) [inline libraries/plugins/es_objects/es_objects.cpp]: I'd recommend not use "asset" as a local variable name, since there is a class or struct has same name. I'm not sure if it will cause trouble.

**Comment by oxarbitrage** [comment]
don't merge yet please, need a few more changes.

**Comment by abitmore** [comment]
This discussion is interesting: https://github.com/steemit/steem/issues/2191

**Comment by oxarbitrage** [comment]
i am having a problem with the limit orders saving. it seems that when limit and fill are in the same block order 1.7.X is  not  created. the market history plugin subscribes to `applied_block`, maybe i that one at least for limit orders. 

> This discussion is interesting: steemit/steem#2191
will check this.

**Comment by abitmore** [comment]
When a new limit order is filled in the same block, it will be created then removed, so you can't get the limit order with `db.get(...)` after the block is applied. The observer mechanism used in #732 may help.

**Comment by pnomolos** [comment]
@oxarbitrage I should now be able to go ahead with https://github.com/bitshares/bitshares-ui/issues/661, right?

**Comment by oxarbitrage** [comment]
@pnomolos close to it but not yet i think. i have a documentation for the plugin i will post in this next days that will make it easier to understand, there is going to probably be a hotfix before the next bitshares-core release for a bug we already identified in the limit orders tracking. i will wait until that to get started.

**Comment by pnomolos** [comment]
@oxarbitrage Thanks!
